db/model/subject.js

```diff
@@ -539,37 +542,122 @@ module.exports = function subject(seq, dataTypes) {
-            value1 = inst.getDataValue(key1);
-            param = 'id';
-            param1 = 'absolutePath';
-          } else {
-            key = 'parentAbsolutePath';
-            value = inst.getDataValue(key);
-            key1 = 'parentId';
-            value1 = inst.getDataValue(key1);
-            param = 'absolutePath';
-            param1 = 'id';
+          /**
+           * Set the values to parentAbsolutePath or parentId
+           * @param {Boolean} bool If true, set to parentAbsolutePath, else
+           * set to parentId
+          */
+          function setValues(bool) {
+            if (bool) {
+              key = 'parentAbsolutePath';
+              value = inst.getDataValue(key);
+              key1 = 'parentId';
+              value1 = inst.getDataValue(key1);
+              param = 'absolutePath';
+              param1 = 'id';
+            } else {
+              key = 'parentId';
+              value = inst.getDataValue(key);
+              key1 = 'parentAbsolutePath';
+              value1 = inst.getDataValue(key1);
+              param = 'id';
+              param1 = 'absolutePath';
+            }
+          }
+
+          /**
+           * If both parentId and parentAbsolutePath changed, check whether
+           * the parents match. If they match, return the parent.
+           * Else throw bad request
+           *
+           * If only one of parentId and parentAbsolutePath changed,
+           * return the parent
+           */
+          function checkParent() {
+            // by default, set params to use parentAbsolutePath
+            setValues(true);
+
+            // when parentAbsolutePath is supplied
+            if (inst.changed('parentAbsolutePath')) {
+              if (inst.parentAbsolutePath) {
+
+                // find subject whose absolutePath === parentAbsolutePath
+                // if it does not exist, throw 404 subject not found
+                // if subject exists, does parent id === parentId?
+                // if true, return parent
+                return new seq.Promise((resolve, reject) => {
+                  if (inst.parentAbsolutePath === inst.absolutePath) {
+                    throw new IllegalSelfParenting({
+                      message: 'absolutePath cannot equal parentAbsolutePath: ' + value,
+                    });
+                  }
+
+                  Subject.findOne({
+                    where: { absolutePath: value },
+                  })
+                  .then((parent) => {
+                    // if parentId changed too, check for consistency
+                    if (inst.changed('parentId') && inst.parentId) {
+
+                      if (parent && parent.id !== value1) {
+                        // parentAbsolutePath does not match parentId
+                        throw new ParentSubjectNotMatch({
+                          message: value + ' does not match ' + value1,
+                        });
+                      } else if (!parent) {
```

Comment:
Can this condition be outside  if (inst.changed('parentId') && inst.parentId) check?